core/store/src/flat_state.rs

```diff
@@ -352,49 +357,130 @@ pub struct FlatStorageState(Arc<RwLock<FlatStorageStateInner>>);
+        store: &Store,
+        shard_id: ShardId,
+        block_hash: CryptoHash,
+    ) -> Result<Option<FlatStateDelta>, crate::StorageError> {
+        let key = KeyForFlatStateDelta { shard_id, block_hash };
+        store
+            .get_ser(crate::DBCol::FlatStateDeltas, &key.try_to_vec().unwrap())
+            .map_err(|_| crate::StorageError::StorageInternalError)
+    }
+
+    pub fn set_delta(
+        store_update: &mut StoreUpdate,
+        shard_id: ShardId,
+        block_hash: CryptoHash,
+        delta: &FlatStateDelta,
+    ) -> Result<(), crate::StorageError> {
+        let key = KeyForFlatStateDelta { shard_id, block_hash };
+        store_update
+            .set_ser(crate::DBCol::FlatStateDeltas, &key.try_to_vec().unwrap(), delta)
+            .map_err(|_| crate::StorageError::StorageInternalError)
+    }
+
+    pub(crate) fn get_flat_head(store: &Store, shard_id: ShardId) -> CryptoHash {
+        store
+            .get_ser(crate::DBCol::FlatStateMisc, &shard_id.try_to_vec().unwrap())
+            .expect("Error reading flat head from storage")
+            .expect(&format!("Cannot read flat head for shard {} from storage", shard_id))
+    }
+
+    pub fn set_flat_head(store_update: &mut StoreUpdate, shard_id: ShardId, val: &CryptoHash) {
+        store_update
+            .set_ser(crate::DBCol::FlatStateMisc, &shard_id.try_to_vec().unwrap(), val)
+            .expect("Error writing flat head from storage")
+    }
+}
+
 impl FlatStorageState {
-    #[allow(unused)]
+    /// Create a new FlatStorageState for `shard_id`.
+    /// The root block of the FlatStorageState is initialized to be what is stored on storage.
+    /// We also load all blocks with height between the root block to the current chain head,
+    /// including those on forks into the returned FlatStorageState.
+    #[cfg(feature = "protocol_feature_flat_state")]
     pub fn new(
         store: Store,
-        head: CryptoHash,
-        tail: CryptoHash,
-        block_headers: HashMap<CryptoHash, BlockHeader>,
+        chain_head_height: BlockHeight,
+        // Unfortunately we don't have access to ChainStore inside this file because of package
+        // dependencies, so we pass these functions in to access chain info
+        hash_to_header: &dyn Fn(&CryptoHash) -> BlockHeader,
+        height_to_hashes: &dyn Fn(BlockHeight) -> HashSet<CryptoHash>,
         shard_id: ShardId,
     ) -> Self {
-        // TODO (#7327): load deltas from storage
-        let deltas = HashMap::new();
+        let flat_head = store_helper::get_flat_head(&store, shard_id);
+        let header = hash_to_header(&flat_head);
+        let flat_head_height = header.height();
+        let mut blocks = HashMap::from([(
+            flat_head,
+            BlockInfo { hash: flat_head, height: flat_head_height, prev_hash: *header.prev_hash() },
+        )]);
+        let mut deltas = HashMap::new();
+        for height in flat_head_height + 1..chain_head_height {
```

Comment:
I might be missing something but I thought `chain_head_height` should also be included here, i.e. I expected the loop range to be inclusive here.

Reply:
Good catch!